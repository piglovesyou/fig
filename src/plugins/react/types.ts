import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';

export type ReactCursorType = NodePath<JSXElement>;
