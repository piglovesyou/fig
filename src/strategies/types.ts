import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { GenContext } from '../gen/make-gen-context';
import {
  ParentVisitContext,
  VisitContext,
  VisitContextWithCursor,
} from '../visit/visit-context';

export interface Strategy {
  makeLayout(): NodePath<JSXElement>;
  postWalk(): void;
  render(): string;
  appendComponentInstanceElement(
    context: VisitContext,
    parentContext: VisitContextWithCursor,
    genContext: GenContext
  ): void;
  appendElement(
    context: VisitContext,
    parentContext: VisitContextWithCursor,
    tagName?: string
  ): NodePath<JSXElement>;
  appendSvgElement(
    context: VisitContext,
    parentContext: ParentVisitContext,
    svgHtml: string
  ): void;
  appendTextElement(
    context: VisitContext,
    parentContext: ParentVisitContext
  ): void;
}
