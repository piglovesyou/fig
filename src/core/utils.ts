import { ComposableNode, Node } from '../types/ast';

export function isValidComponentNode(
  node: ComposableNode
): node is Node<'FRAME'> | Node<'COMPONENT'> | Node<'INSTANCE'> {
  return (
    (node.type === 'FRAME' ||
      node.type === 'COMPONENT' ||
      node.type === 'INSTANCE') &&
    node.visible !== false
  );
}

export function walkNodeTree<T = void>(
  node: ComposableNode,
  fn: (node: ComposableNode, parentResult: T) => null | T,
  parentResult: T
): void {
  const result = fn(node, parentResult);
  if (result === null) return;
  for (const child of node.children || []) {
    walkNodeTree(child, fn, result);
  }
}
