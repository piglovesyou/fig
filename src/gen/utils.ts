import camelCase from 'camelcase';
import { ComposableNode } from '../types/ast';

function makeComponentNameFromId(id: string, name: string) {
  id = id.replace(/;/g, '$$').replace(/\W/g, '$');
  name = camelCase(name, { pascalCase: true });
  return `${name}_${id}`;
}

export function makeComponentName(node: ComposableNode): string | never {
  switch (node.type) {
    case 'FRAME':
    case 'COMPONENT':
      return makeComponentNameFromId(node.id, node.name);
    case 'INSTANCE':
      // TODO: zombie component name can be resolved. Fix it.
      return makeComponentNameFromId(node.id, '__NOT_FOUND__');
    default:
      throw new Error('Component can only be FRAME, COMPONENT, INSTANCE.');
  }
}
