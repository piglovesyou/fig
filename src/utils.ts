import camelCase from 'camelcase';
import _ from 'lodash';
import { Node } from './types';
import {
  Color,
  ComposableNode,
  EffectShadow,
  Paint,
  PaintSolidScaleMode,
} from './types/ast';

// const vectorTypes = [
//   'VECTOR',
//   'LINE',
//   'REGULAR_POLYGON',
//   'ELLIPSE',
//   'STAR',
//   'SHAPE',
// ];

// export function isVector(node: ComposableNode) {
//   return vectorTypes.includes(node.type);
// }

// const GROUP_TYPES = ['GROUP', 'BOOLEAN_OPERATION'];

function expand256(color: number) {
  return Math.round(color * 255);
}

function toColorHex(c: number): string {
  const s = c.toString(16);
  if (s.length === 1) return `${s}${s}`;
  return s;
}

export function colorString(color?: Color) {
  if (!color) {
    console.warn('!!!color can be undefined!!!');
    return 'transparent';
  }
  const r = expand256(color.r);
  const g = expand256(color.g);
  const b = expand256(color.b);
  if (color.a === 1) {
    return `#${toColorHex(r)}${toColorHex(g)}${toColorHex(b)}`;
  }
  return `rgba(${r}, ${g}, ${b}, ${color.a})`;
}

export function dropShadow(effect: EffectShadow) {
  return `${effect.offset.x}px ${effect.offset.y}px ${
    effect.radius
  }px ${colorString(effect.color)}`;
}

export function innerShadow(effect: EffectShadow) {
  return `inset ${effect.offset.x}px ${effect.offset.y}px ${
    effect.radius
  }px ${colorString(effect.color)}`;
}

export function imageURL(hash: string) {
  const squash = hash.split('-').join('');
  return `url(https://s3-us-west-2.amazonaws.com/figma-alpha/img/${squash.substring(
    0,
    4
  )}/${squash.substring(4, 8)}/${squash.substring(8)})`;
}

export function backgroundSize(scaleMode: PaintSolidScaleMode) {
  if (scaleMode === 'FILL') {
    return 'cover';
  }
}

export function sortByYAxis(a: ComposableNode, b: ComposableNode): number {
  return a.absoluteBoundingBox.y - b.absoluteBoundingBox.y;
}

export function getLastVisible<T>(arra: T[]): T | null {
  // @ts-ignore
  return _.findLast(arra, (e: any) => e.visible != false);
}

export function paintToLinearGradient(paint: Paint) {
  const handles = paint.gradientHandlePositions || [];
  const handle0 = handles[0];
  const handle1 = handles[1];

  const ydiff = handle1.y - handle0.y;
  const xdiff = handle0.x - handle1.x;

  const angle = Math.atan2(-xdiff, -ydiff);
  const stops = (paint.gradientStops || [])
    .map((stop) => {
      return `${colorString(stop.color)} ${Math.round(stop.position * 100)}%`;
    })
    .join(', ');
  return `linear-gradient(${angle}rad, ${stops})`;
}

export function paintToRadialGradient(paint: Paint) {
  const stops = (paint.gradientStops || [])
    .map((stop) => {
      return `${colorString(stop.color)} ${Math.round(stop.position * 60)}%`;
    })
    .join(', ');

  return `radial-gradient(${stops})`;
}

export function makeComponentNameFromId(id: string, name: string) {
  id = id.replace(/;/g, '$$').replace(/\W/g, '$');
  name = camelCase(name, { pascalCase: true });
  return `${name}_${id}`;
}

export function makeComponentName(node: ComposableNode): string | never {
  switch (node.type) {
    case 'FRAME':
    case 'COMPONENT':
      return makeComponentNameFromId(node.id, node.name);
    case 'INSTANCE':
      // TODO: zombie component name can be resolved. Fix it.
      return makeComponentNameFromId(node.id, '__NOT_FOUND__');
    default:
      throw new Error('Component can only be FRAME, COMPONENT, INSTANCE.');
  }
}

export type ComponentInfo = {
  name: string;
  node: ComposableNode;
};
export type ComponentsMap = Map<string, ComponentInfo>;

export function isValidComponentNode(
  node: ComposableNode
): node is Node<'FRAME'> | Node<'COMPONENT'> | Node<'INSTANCE'> {
  return (
    (node.type === 'FRAME' ||
      node.type === 'COMPONENT' ||
      node.type === 'INSTANCE') &&
    node.visible !== false
  );
}

export function walkNodeTree(
  node: ComposableNode,
  fn: (node: ComposableNode) => void | boolean
): void {
  const result = fn(node);
  if (result === false) return;
  for (const child of node.children || []) {
    walkNodeTree(child, fn);
  }
}
