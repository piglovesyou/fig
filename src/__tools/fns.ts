import { join } from 'path';
import { renderToStaticMarkup } from 'react-dom/server';
import { readFile } from '../utils/fs';

export function renderInHtml(element: JSX.Element) {
  return `<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no" />
<style>
  
body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
  position: absolute;
  width: 100vw;
  min-height: 100vh;
}

body > * {
  overflow: hidden;
  min-width: 100vw;
  min-height: 100vh;
}

/*.vector svg {*/
/*  width: 100%;*/
/*  height: 100%;*/
/*  position: absolute;*/
/*}*/

/*a {*/
/*  cursor: pointer;*/
/*}*/

</style>
</head>
<body>${renderToStaticMarkup(element)}</body>
</html>`;
}

export async function readJson(...paths: string[]) {
  const fullPath = join(...paths);
  return JSON.parse(await readFile(fullPath, 'utf-8'));
}
