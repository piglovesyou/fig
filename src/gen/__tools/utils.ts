import generate from '@babel/generator';
import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { format } from 'prettier';
import { createConfig } from '../../core/config';
import { ComposableNode, FigmaFile } from '../../types/fig';
import { readFile } from '../../utils/fs';
import { makeGenContext } from '../make-gen-context';

export async function makeCreateComponentArgs(
  fileJsonFullPath: string,
  fileKey: string,
  baseDir: string
) {
  const figmaFileFixture = JSON.parse(
    await readFile(fileJsonFullPath, 'utf-8')
  );
  const node: ComposableNode =
    figmaFileFixture.document.children[0].children[0];
  expect(node.name).toBe('My Frame');

  const config = await createConfig({
    fileKeys: [fileKey],
    baseDir,
    plugins: ['react'],
  });
  const genContext = await makeGenContext(
    figmaFileFixture as FigmaFile,
    fileKey,
    config,
    baseDir
  );

  const { componentsMap } = genContext;
  const componentInfo = componentsMap.get('1:2');
  if (!componentInfo) throw new Error('Never');
  return [node, componentInfo, genContext] as const;
}

export function getCode(rootCursor: NodePath<JSXElement>) {
  const { code } = generate(rootCursor.node);
  return format(code, { parser: 'babel' });
}
