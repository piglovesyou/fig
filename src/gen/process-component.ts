import { NodePath } from '@babel/traverse';
import { File, JSXElement } from '@babel/types';
import { readFile, writeFile } from 'fs/promises';
import makeDir from 'make-dir';
import { join } from 'path';
import { isValidComponentNode, walkNodeTree } from '../core/node-utils';
import { JsxStrategy } from '../strategies/jsx';
import { parseAsRoot } from '../strategies/jsx/jsx-utils';
import { visitNode } from '../visit/visit';
import { EmptyVisitContext } from '../visit/visit-context';
import { ComponentInfo } from './components-map';
import { GenContext } from './make-gen-context';

export async function processComponent(
  componentInfo: ComponentInfo,
  genContext: GenContext
) {
  const { node, name } = componentInfo;
  if (!isValidComponentNode(node)) throw new Error('never');

  const strategy = new JsxStrategy(componentInfo);

  const { baseFullDir } = genContext;
  const componetsDir = join(
    baseFullDir,
    node.type === 'FRAME'
      ? genContext.config.pagesDir
      : genContext.config.componentsDir
  );
  await makeDir(componetsDir);
  const fullPath = join(componetsDir, `${name}.tsx`);

  let rootAst: File;
  let placeholderCursor: NodePath<JSXElement>;

  if (false /*existsSync(fullPath)*/) {
    // TODO: Update mode.
    // 1. Parse and get the File ast
    // 2. Find the root element to update
    // 3. Repeat update and traverse
    const content = await readFile(fullPath, 'utf-8');
    rootAst = parseAsRoot(content);
    throw new Error('Implement');
  } else {
    // Create mode.
    placeholderCursor = strategy.makeLayout();

    const parentContext: EmptyVisitContext = { cursor: placeholderCursor };
    walkNodeTree(
      node,
      (node, parentContext) => {
        return visitNode(node, parentContext, strategy, genContext);
      },
      parentContext
    );
    strategy.postWalk();
  }

  await writeFile(fullPath, strategy.render());
}

// // TODO: Similar to preprocess. Refactor
// export async function processCanvas(
//   canvas: Canvas,
//   genContext: GenContext
// ): Promise<void> {
//   for (const screen of canvas.children) {
//     if (isValidComponentNode(screen))
//       await processComponent(screen, genContext);
//   }
// }
