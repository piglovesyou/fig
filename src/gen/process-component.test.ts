import { NodePath } from '@babel/traverse';
import { JSXElement } from '@babel/types';
import { join } from 'path';
import { createComponent } from './process-component';
import { getCode, makeCreateComponentArgs } from './__tools/utils';

const fileJsonFullPath = join(
  __dirname,
  './__fixtures/YiawDxsoqZYkiufOizx367.file.json'
);
const fileKey = process.env.TOKEN!;
const baseDir = join(__dirname, '__generated__');

describe('process-component.test.ts', () => {
  test('createComponent', async () => {
    const args = await makeCreateComponentArgs(
      fileJsonFullPath,
      fileKey,
      baseDir
    );
    const rootCursor = await createComponent<NodePath<JSXElement>>(...args);

    const code = getCode(rootCursor);
    expect(code).toMatchSnapshot();
  });
});
