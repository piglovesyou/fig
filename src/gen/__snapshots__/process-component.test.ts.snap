// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process-component.test.ts createComponent 1`] = `
"<__PLACEHOLDER__>
  <div
    data-fid=\\"1:2\\"
    style={{
      ...{
        display: \\"flex\\",
        flexDirection: \\"column\\",
        paddingTop: 16,
        paddingRight: 16,
        paddingBottom: 16,
        paddingLeft: 16,
        gap: 16,
        position: \\"absolute\\",
        boxSizing: \\"border-box\\",
        backgroundColor: \\"#ffffff\\",
      },
      ...props.style,
    }}
    data-fname=\\"My Frame\\"
  >
    <MyComponent_4$4
      data-fid=\\"4:5\\"
      style={{
        display: \\"flex\\",
        flexDirection: \\"row\\",
        gap: 4,
        alignItems: \\"center\\",
        position: \\"relative\\",
        boxSizing: \\"border-box\\",
        flexGrow: 0,
        alignSelf: \\"inherit\\",
        width: 139,
        height: 24,
        backgroundColor: \\"rgba(0, 0, 0, 0)\\",
      }}
      data-fname=\\"My Component\\"
    ></MyComponent_4$4>
    <MyComponent_4$4
      data-fid=\\"4:8\\"
      style={{
        display: \\"flex\\",
        flexDirection: \\"row\\",
        gap: 4,
        alignItems: \\"center\\",
        position: \\"relative\\",
        boxSizing: \\"border-box\\",
        flexGrow: 0,
        alignSelf: \\"inherit\\",
        width: 139,
        height: 24,
        backgroundColor: \\"rgba(0, 0, 0, 0)\\",
      }}
      data-fname=\\"My Component\\"
    ></MyComponent_4$4>
    <MyComponent_4$4
      data-fid=\\"4:11\\"
      style={{
        display: \\"flex\\",
        flexDirection: \\"row\\",
        gap: 4,
        alignItems: \\"center\\",
        position: \\"relative\\",
        boxSizing: \\"border-box\\",
        flexGrow: 0,
        alignSelf: \\"inherit\\",
        width: 139,
        height: 24,
        backgroundColor: \\"rgba(0, 0, 0, 0)\\",
      }}
      data-fname=\\"My Component\\"
    ></MyComponent_4$4>
  </div>
</__PLACEHOLDER__>;
"
`;
